BaragonService/src/main/java/com/hubspot/baragon/service/worker/BaragonRequestWorker.java

```diff
@@ -327,23 +361,74 @@ private QueuedRequestWithState setNoValidateIfRequestRemovesUpstreamsOnly(Queued
           .setNoDuplicateUpstreams(originalRequest.isNoDuplicateUpstreams())
           .build();
 
-      try {
-        requestManager.updateRequest(requestWithNoValidate);
-      } catch (Exception e) {
-        // This is just an optimization, so don't blow up if it fails.
-        LOG.warn("Unable to set noValidate for request {}", nonServiceChangeRequest.getQueuedRequestId().getRequestId(), e);
-      }
-
-      return new QueuedRequestWithState(
-          nonServiceChangeRequest.getQueuedRequestId(),
+      return new MaybeAdjustedRequest(new QueuedRequestWithState(
+          nonServiceChangeRequest.request.getQueuedRequestId(),
           requestWithNoValidate,
-          nonServiceChangeRequest.getCurrentState()
-      );
+          nonServiceChangeRequest.request.getCurrentState()
+      ), true);
+    }
+
+    return nonServiceChangeRequest;
+  }
+
+  private MaybeAdjustedRequest preResolveDNS(MaybeAdjustedRequest nonServiceChangeRequest) {
+    if (!nonServiceChangeRequest.request.getRequest().getLoadBalancerService().isPreResolveUpstreamDNS()) {
+      return nonServiceChangeRequest;
+    }
+
+    BaragonRequest originalRequest = nonServiceChangeRequest.request.getRequest();
+
+    List<UpstreamInfo> maybeResolvedAddUpstreams = resolveDNSForAllUpstreams(nonServiceChangeRequest.request.getRequest().getAddUpstreams());
+    List<UpstreamInfo> maybeResolvedRemoveUpstreams = resolveDNSForAllUpstreams(nonServiceChangeRequest.request.getRequest().getRemoveUpstreams());
+    List<UpstreamInfo> maybeResolvedReplaceUpstreams = resolveDNSForAllUpstreams(nonServiceChangeRequest.request.getRequest().getReplaceUpstreams());
+
+    if (allUpstreamsAreResolved(maybeResolvedAddUpstreams)
+        && allUpstreamsAreResolved(maybeResolvedRemoveUpstreams)
+        && allUpstreamsAreResolved(maybeResolvedReplaceUpstreams)) {
+      LOG.trace("Request {} does not change a BaragonService and all upstreams were pre-resolved. Setting noValidate.", nonServiceChangeRequest.request.getQueuedRequestId().getRequestId());
+      BaragonRequest requestWithResolvedUpstreams = new BaragonRequestBuilder()
```

Comment:
is it worth making a toBuilder on BaragonRequest?

Reply:
Probs, yeah. There's a lot of copy boilerplate at the moment